### PR TITLE
feat: Update OpenAI spec to include image url in message content

### LINKED ISSUE #107: Add support to include images for OpenAI Chat Template
### Description

Add support for images to be included in chat messages, similar to [gpt-4o](https://platform.openai.com/docs/guides/vision).

```json
{
  "messages": [
     {
            "role": "user",
            "content": [
                {"type": "text", "text": "What's in this image?"},
                {
                    "type": "image_url",
                    "image_url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
                },
            ],
        }
  ]
}
```

### Motivation

I was trying to implement a chat generation API using the phi3-vision model, which takes images as input. The proposed format will facilitate interaction with models that involve images. This approach will simplify the handling of different prompt techniques, which can be managed separately before passing the prompt to the model.

### Benefits

- **Enhanced Interaction**: Enables richer interactions by allowing images in chat messages.
- **Flexibility**: Provides a structured way to handle various prompt techniques involving images.

### Additional Context

Implementing this feature will involve updating the schema to support image URLs in chat messages, ensuring compatibility with existing and future vision models.

### PR DESCRIPTION
**Pull request by bhimrazy**
<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/pytorch-lightning/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #107.
- Add support for images to be included in chat messages, similar to [gpt-4o](https://platform.openai.com/docs/guides/vision).

### CODE REVIEW

**Comment by lantiga** [comment]
Thanks for the PR @bhimrazy!
Can you add a test and a minimal end to end example on the readme?

**Comment by williamFalcon** [comment]
@bhimrazy sick! super excited to try this.

@lantiga @lantiga do we have a guide or something to show how to add the test and example?

**Comment by lantiga** [comment]
no, good point /cc @aniketmaurya 

@bhimrazy for now you can take inspiration from:

- https://github.com/Lightning-AI/LitServe/blob/main/tests/test_specs.py for unit tests
- https://github.com/Lightning-AI/LitServe/blob/main/tests/e2e/test_e2e.py#L90 for the end to end test
- https://github.com/Lightning-AI/LitServe/blob/main/README.md?plain=1#L548 for the README examples

**Comment by aniketmaurya** [comment]
thank you for the PR @bhimrazy! as Luca mentioned, you can take inspiration from the existing [LitSpec test cases](https://github.com/Lightning-AI/LitServe/blob/main/tests/test_specs.py). 


Maybe you can try sending the request with image content to the server and check that it is able to parse and doesn't break.

```
{
  "model": "lit",
  "messages": [
     {
            "role": "user",
            "content": [
                {"type": "text", "text": "What's in this image?"},
                {
                    "type": "image_url",
                    "image_url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
                },
            ],
        }
  ]
}
}
```

**Comment by bhimrazy** [comment]
Thanks, @lantiga, @williamFalcon, and @aniketmaurya for all of this feedback. 
I will go through the given examples and add the test cases and e2e example.

**Comment by aniketmaurya** [comment]
awesome @bhimrazy!! don't hesitate to reach out if you need any help.

**Comment by bhimrazy** [comment]
Hi @aniketmaurya @lantiga , Could you please help me through the addition of end-to-end documentation to the README? 
I have prepared a draft version of it, included below.

---

LitServe's OpenAISpec also enables capability to handle images in the input. Below is an example of how to set this up using LitServe.

```python
import litserve as ls
from litserve.specs.openai import ChatMessage

class OpenAISpecLitAPI(ls.LitAPI):
    def setup(self, device):
        self.model = None

    def predict(self, x):
        yield {"role": "assistant", "content": "This is a generated output"}

    def encode_response(self, output: dict) -> ChatMessage:
        yield ChatMessage(role="assistant", content="This is a custom encoded output")


if __name__ == "__main__":
    server = ls.LitServer(OpenAISpecLitAPI(), spec=ls.OpenAISpec())
    server.run(port=8000)
 ```

 In this case, `predict` is expected to take an input with the following shape:
- Text Input Example:
    ```
    [{"role": "system", "content": "You are a helpful assistant."},
     {"role": "user", "content": "Hello there"},
     {"role": "assistant", "content": "Hello, how can I help?"},
     {"role": "user", "content": "What is the capital of Australia?"}]
    ```
- Mixed Text and Image Input Example:

    ```
    [{"role": "system", "content": "You are a helpful assistant."},
     {
     "role": "user", 
     "content": [
                    {"type": "text", "text": "What's in this image?"},
                    {
                        "type": "image_url",
                        "image_url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
                    },
                ]
    
    },
     {"role": "assistant", "content": "A wooden boardwalk through a green field under a blue sky."},
     {"role": "user", "content": "How is the weather depicted in the image?"}]
    ```

The above server can be queried using a standard OpenAI client:
```python
import requests

response = requests.post("http://127.0.0.1:8000/v1/chat/completions", json={
    "model": "my-gpt2",
    "stream": False,  # You can stream chunked response by setting this True
    "messages": [
      {
        "role": "system",
        "content": "You are a helpful assistant."
      },
      {
        "role": "user",
        "content": [
                {"type": "text", "text": "What's in this image?"},
                {
                    "type": "image_url",
                    "image_url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
                },
            ]
      }
    ]
  })
```

**Comment by aniketmaurya** [comment]
looks good @bhimrazy! would be nice if you can show that image content could be processed in predict or decode_request step. For example:

```
    def predict(self, x):
        if isinstance(x["content"], list):
                image_caption = self.model(x["content"][1]["image_url"])
                yield {"role": "assistant", "content": image_caption}
        else:
            yield {"role": "assistant", "content": "This is a generated output."}
```

**Comment by bhimrazy** [comment]
> looks good @bhimrazy! would be nice if you can show that image content could be processed in predict or decode_request step. For example:
> 
> ```python
>     def predict(self, x):
>         if isinstance(x["content"], list):
>                 # do something with image url
>                 image_url = x["content"][1]["image_url"]
>                 yield {"role": "assistant", "content": "the image describes nature and bla bla..."}
>         else:
>             yield {"role": "assistant", "content": "This is a generated output."}
> ```

Sure Thanks!

File: src/litserve/specs/openai.py [hunk 1]
```diff
@@ -42,9 +42,19 @@ class UsageInfo(BaseModel):
     completion_tokens: Optional[int] = 0
 
 
+class TextContentPart(BaseModel):
```
Comment (lantiga) [inline src/litserve/specs/openai.py:45]:
```suggestion
class TextContent(BaseModel):
```

File: src/litserve/specs/openai.py [hunk 2]
```diff
@@ -42,9 +42,19 @@ class UsageInfo(BaseModel):
...
+    type: str
+    text: str
+
+
+class ImageContentPart(BaseModel):
```
Comment (lantiga) [inline src/litserve/specs/openai.py:50]:
```suggestion
class ImageContent(BaseModel):
```

File: src/litserve/specs/openai.py [hunk 3]
```diff
@@ -42,9 +42,19 @@ class UsageInfo(BaseModel):
...
+    type: str
+    image_url: str
+
+
 class ChatMessage(BaseModel):
     role: str
-    content: str
+    content: Union[str, List[Union[TextContentPart, ImageContentPart]]]
```
Comment (lantiga) [inline src/litserve/specs/openai.py:57]:
```suggestion
    content: Union[str, List[Union[TextContent, ImageContent]]]
```

File: README.md [hunk 4]
```diff
@@ -640,7 +640,90 @@ if __name__ == "__main__":
     server = ls.LitServer(GPT2LitAPI(), spec=ls.OpenAISpec())
     server.run(port=8000)
 ```
+&nbsp;
+
+LitServe's OpenAISpec also enables capability to handle images in the input. Below is an example of how to set this up using LitServe.
```
Comment (lantiga) [inline README.md:645]:
```suggestion
LitServe's `OpenAISpec` can also handle images in the input. Here is an example:
```

File: README.md [hunk 5]
```diff
@@ -640,7 +640,90 @@ if __name__ == "__main__":
...
+    def encode_response(self, output: dict) -> ChatMessage:
+        yield ChatMessage(role="assistant", content="This is a custom encoded output")
+
+
+if __name__ == "__main__":
+    server = ls.LitServer(OpenAISpecLitAPI(), spec=ls.OpenAISpec())
+    server.run(port=8000)
+```
+
+In this case, `predict` is expected to take an input with the following shape:
```
Comment (lantiga) [inline README.md:673]:
```suggestion
When using `OpenAISpec`, `predict` is expected to take an input with the following shape:
```

File: README.md [hunk 6]
```diff
@@ -640,7 +640,90 @@ if __name__ == "__main__":
...
+                  {"type": "text", "text": "What's in this image?"},
+                  {
+                      "type": "image_url",
+                      "image_url": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/dd/Gfp-wisconsin-madison-the-nature-boardwalk.jpg/2560px-Gfp-wisconsin-madison-the-nature-boardwalk.jpg",
+                  },
+              ]
+
+  },
+   {"role": "assistant", "content": "This image shows a wooden boardwalk extending through a lush green marshland."},
+   {"role": "user", "content": "How is the weather depicted in the image?"}]
```
Comment (lantiga) [inline README.md:698]:
```suggestion
   {"role": "user", "content": "What is the weather like in the image?"}]
```

**Review by lantiga — APPROVED** [review]
Looks great!

**Comment by lantiga** [comment]
Awesome job @bhimrazy, let's see what CI thinks and then we're ready to merge!

**Comment by williamFalcon** [comment]
congrats @bhimrazy!
solid contribution

**Pull request merged**